File: problem101.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool IsSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return IsSymmetricHelper(root.left, root.right);
    }

    private bool IsSymmetricHelper(TreeNode node1, TreeNode node2) {
        if (node1 == null && node2 == null) {
            return true;
        } else if (node1 == null || node2 == null) {
            return false;
        }

        return node1.val == node2.val 
            && IsSymmetricHelper(node1.left, node2.right)
            && IsSymmetricHelper(node1.right, node2.left);
    }
}

File: su19/problem101.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        return self.checkBranches(root.left, root.right)

    def checkBranches(self, left, right):
        if left == None and right == None:
            return True
        if left == None or right == None:
            return False
        if (left.val != right.val):
            return False
        return self.checkBranches(left.right, right.left) \
            and self.checkBranches(left.left, right.right)